pkg/frontend/querymiddleware/read_consistency_test.go

```diff
@@ -0,0 +1,136 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package querymiddleware
+
+import (
+	"context"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/grafana/dskit/flagext"
+	"github.com/grafana/dskit/services"
+	"github.com/grafana/dskit/user"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"github.com/twmb/franz-go/pkg/kgo"
+
+	querierapi "github.com/grafana/mimir/pkg/querier/api"
+	"github.com/grafana/mimir/pkg/storage/ingest"
+	"github.com/grafana/mimir/pkg/util/testkafka"
+)
+
+func TestReadConsistencyRoundTripper(t *testing.T) {
+	const (
+		topic         = "test"
+		numPartitions = 10
+		tenantID      = "user-1"
+	)
+
+	tests := map[string]struct {
+		limits          Limits
+		reqConsistency  string
+		expectedOffsets bool
+	}{
+		"should not inject offsets if default read consistency is 'eventual' and request has explicitly requested any consistency level": {
+			limits:          mockLimits{ingestStorageReadConsistency: querierapi.ReadConsistencyEventual},
+			expectedOffsets: false,
+		},
+		"should not inject offsets if default read consistency is 'strong' and request has explicitly requested 'eventual' consistency": {
+			limits:          mockLimits{ingestStorageReadConsistency: querierapi.ReadConsistencyStrong},
+			reqConsistency:  querierapi.ReadConsistencyEventual,
+			expectedOffsets: false,
+		},
+		"should inject offsets if default read consistency is 'eventual' but request has explicitly requested 'strong' consistency": {
+			limits:          mockLimits{ingestStorageReadConsistency: querierapi.ReadConsistencyEventual},
+			reqConsistency:  querierapi.ReadConsistencyStrong,
+			expectedOffsets: true,
+		},
+		"should inject offsets if default read consistency is 'strong' and request has not explicitly requested any consistency level": {
+			limits:          mockLimits{ingestStorageReadConsistency: querierapi.ReadConsistencyStrong},
+			expectedOffsets: true,
+		},
+	}
+
+	for testName, testData := range tests {
+		t.Run(testName, func(t *testing.T) {
+			// Capture the downstream HTTP request.
+			var downstreamReq *http.Request
+			downstream := RoundTripFunc(func(req *http.Request) (*http.Response, error) {
+				downstreamReq = req
+				return nil, nil
+			})
+
+			ctx := context.Background()
+			logger := log.NewNopLogger()
+
+			_, clusterAddr := testkafka.CreateCluster(t, numPartitions, topic)
```

Comment:
a fully-fledged kafka looks like an overkill for this test. `TopicOffsetsReader` is already extensively tested in the `ingest` package. WDYT about mocking the `NewTopicOffsetsReader`?

Reply:
I personally prefer testing it more end-to-end rather than mocking `TopicOffsetsReader`. Using `kfake` and the real `TopicOffsetsReader` increases my confidence.